tests/integration/test_sqs.py

```diff
@@ -1194,6 +1194,390 @@ def receive_and_check_order():
+
+        # even though we sent a message from group 2 in between messages of group 1, the messages arrive in
+        # order of the message groups
+        assert response["Messages"][0]["Body"] == "g1-m1"
+        assert response["Messages"][1]["Body"] == "g1-m2"
+        assert response["Messages"][2]["Body"] == "g1-m3"
+        assert response["Messages"][3]["Body"] == "g2-m1"
+
+    @pytest.mark.aws_validated
+    def test_fifo_receive_message_visibility_timeout_shared_in_group(
+        self, sqs_create_queue, aws_client
+    ):
+        timeout = 1
+        queue_url = sqs_create_queue(
+            QueueName=f"queue-{short_uid()}.fifo",
+            Attributes={
+                "FifoQueue": "true",
+                "VisibilityTimeout": f"{timeout}",
+                "ContentBasedDeduplication": "true",
+            },
+        )
+
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g1-m1", MessageGroupId="group-1"
+        )
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g1-m2", MessageGroupId="group-1"
+        )
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g1-m3", MessageGroupId="group-1"
+        )
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g1-m4", MessageGroupId="group-1"
+        )
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g2-m1", MessageGroupId="group-2"
+        )
+        aws_client.sqs.send_message(
+            QueueUrl=queue_url, MessageBody="g3-m1", MessageGroupId="group-3"
+        )
+
+        # we receive 2 messages of the 3 in the first message group
+        response = aws_client.sqs.receive_message(QueueUrl=queue_url, MaxNumberOfMessages=2)
+        assert response["Messages"][0]["Body"] == "g1-m1"
+        assert response["Messages"][1]["Body"] == "g1-m2"
+
+        # the entire group 1 is affected by the visibility timeout, so the next receive returns a message
+        # of group 2
+        response = aws_client.sqs.receive_message(QueueUrl=queue_url, MaxNumberOfMessages=1)
+        assert response["Messages"][0]["Body"] == "g2-m1"
+
+        # let the visibility timeout expire
+        time.sleep(timeout + 1)
+
+        # now we try to fetch all messages from the queue. since there are no guarantees about the ordering of
+        # message groups themselves, multiple orderings are valid now
+        response = aws_client.sqs.receive_message(
+            QueueUrl=queue_url, MaxNumberOfMessages=10, VisibilityTimeout=0
+        )
+        ordering = [message["Body"] for message in response["Messages"]]
+        # these are in principle all valid orderings
+        assert ordering in (
+            ["g3-m1", "g1-m1", "g1-m2", "g1-m3", "g1-m4", "g2-m1"],
+            ["g3-m1", "g2-m1", "g1-m1", "g1-m2", "g1-m3", "g1-m4"],
+            ["g1-m1", "g1-m2", "g1-m3", "g1-m4", "g2-m1", "g3-m1"],
+        )
```

Comment:
I'm glad you pointed this out, I was a bit sloppy here. Instead of adding all orderings, I instead changed all the tests assert to be much more specific now. Some test cases were actually the same in AWS and LocalStack. There was just one case where I couldn't at all reproduce how AWS does the ordering of groups, and another where there's clearly a race and two orderings are possible. I also added comments whether the ordering is related to AWS or LocalStack. I think this is much clearer.